zarr/core.py

```diff
@@ -1952,7 +1952,7 @@ def _process_for_setitem(self, ckey, chunk_selection, value, fields=None):
         return self._encode_chunk(chunk)
 
     def _chunk_key(self, chunk_coords):
-        return self._key_prefix + '.'.join(map(str, chunk_coords))
+        return self._key_prefix + self._dimension_separator.join(map(str, chunk_coords))
```

Comment:
Can you explain how this change is compatible with `FSStore._normalize_key`? https://github.com/zarr-developers/zarr-python/blob/master/zarr/storage.py#L1076

`FSStore._normalize_key` assumes that all chunk keys are formatted `foo/bar/0.0.0` -- this assumption is the basis of splitting the chunk key into a prefix and a chunk ID via `key.split('/')`. As I understand it, this change breaks this assumption.

Reply:
Reading the `flat` and `nested` fixtures from this repo (`zarr.open(f"file:///tmp/{x}")[:]`) with some sloppy debugging in place shows:

```
|               | Array._chunk_key | FSStore._normalize_keys |
|---------------|------------------|-------------------------|
| master:nested |  (0, 0) --> 0.0  |       0.0 --> 0.0       |
| master:flat   |  (0, 0) --> 0.0  |       0.0 --> 0.0       |
| PR:nested     |  (0, 0) --> 0/0  |       0/0 --> 0/0       |
| PR:flat       |  (0, 0) --> 0.0  |       0.0 --> 0.0       |
```

which likely points to some logic in FSStore being ripe for removal since the Store is basically just accepting what what the Array has detected. Now, how it is that that's working with _your_ PR, I still haven't figured out.

Reply:
Right, as your test shows this is fine for FSStore (and maybe we don't need this code in the store at all if the chunk keys come pre-normalized). But this situation is dire for N5Stores, which need to be able to re-order the chunk keys before writing to storage.